ft_prepare_sourcemodel.m

```diff
@@ -642,6 +646,37 @@
       % copy the boolean fields over
       sourcemodel = copyfields(mnigrid, sourcemodel, booleanfields(mnigrid));
     end
+
+  case 'basedoncentroids'
+    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+    % compute the centroids of each element of a volumetric mesh and use
+    % them as sources. assume all inside.
+    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+    % this is a model with hexaheders or tetraheders
+    if isfield(headmodel, 'tet')
+      % the subsequent code works both for tetraheders or hexaheders, but assumes the volume elements to be called "hex"
+      headmodel.hex = headmodel.tet;
+      headmodel = rmfield(headmodel, 'tet');
+    end
+
+    % determine the size of the relevant elements
+    numhex = length(headmodel.hex);
+    numelem = length(headmodel.hex(1,:));
+
+    sourcemodel.pos = zeros(numhex,3);
+
+    for i=1:length(headmodel.hex)
```

Comment:
rather than length, better use size(..., 1).

If there is only one hexaheder, the length would return 8, which is not what you expect.